[jit][ir] Add pretty printer for JIT IR

driazati opened the pull request:
Adds some pretty-printing capability to the IR graph to make debugging easier/more human readable, see `torch/csrc/jit/test_jit.cpp:925` and onwards for example outputs. Results aren't perfect yet but it's a start.

Commit:
Add pretty printer for JIT IR

Differential Revision: https://phabricator.intern.facebook.com/D9189808

goldsborough commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +250,175 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    if (scope->hasConstant(unique)) {
+      out << scope->getConstant(unique);
+    } else {
+      throw std::runtime_error("const not found for : " + unique);
```
Please use `AT_ERROR`, it will include a backtrace in the error message

goldsborough commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +250,175 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+std::ostream& prettyPrintInputs(std::ostream & out, const Node* node) {
+  out << "(";
+  auto delimiter = "";
+  for (auto in_value : node->inputs()) {
```
Nit:  prefer `const auto&` or `const auto*`, right now it's hard to know if the iterated thing is a pointer or a value

goldsborough commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +250,175 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+
+    // Offset to skip loop trip count
+    int count = 1;
+    std::map<std::string, Value*> looks;
```
This variable is unused?

goldsborough commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +250,175 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  // if there are subblocks on this node, visit them
+  switch (node->kind()) {
+  case prim::Loop: {
+    out << "while ";
```
I feel like these huge switch cases make it *extremely* easy to forget the `break` at the end. Maybe put each into a little helper function? Or some other solution to this problem

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -110,6 +111,8 @@ struct Scope {
   Scope* parent_;
   Symbol name_;
   std::vector<std::unique_ptr<Scope> > children_;
+  std::map<std::string, int> constants_;
```
Are you very sure you want sorted map? `std::map` is a heavy, sorted RB tree. We always prefer `std::unordered_map` (hashtable) if the sorting is not required

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -139,6 +142,24 @@ struct Scope {
     }
     return current;
   }
+  bool hasConstant(std::string name) {
```
const

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -139,6 +142,24 @@ struct Scope {
+  bool hasConstant(std::string name) {
+    return constants_.find(name) != constants_.end();
+  }
+  int getConstant(std::string name) {
```
`const` (`noexcept`)

goldsborough replied:
Are you sure you want the 0 to be returned when the name is not already in the map?

goldsborough replied:
Otherwise use `map.at(name)` which also has a const version

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -139,6 +142,24 @@ struct Scope {
+  bool hasNameAlias(std::string alias) {
+    return input_aliases_.find(alias) != input_aliases_.end();
+  }
+  std::string getNameAlias(std::string alias) {
```
const

goldsborough replied:
You're taking the `std::string` by value and making a copy each time. I think you want `const std::string&`!

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -139,6 +142,24 @@ struct Scope {
+  int getConstant(std::string name) {
+    return constants_[name];
+  }
+  void addConstant(std::string name, int value) {
```
Copying the string here

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -198,12 +220,25 @@ struct Value {
       return unique_name_;
     return std::to_string(unique());
   }
+  std::string originalName() const {
```
Return `const std::string&`, no need to copy each time

goldsborough commented on torch/csrc/jit/ir.h:
```diff
@@ -198,12 +220,25 @@ struct Value {
+  std::string originalName() const {
+   return original_name_;
+ }
+ std::string readableName() const {
```
Same here

goldsborough reviewed (commented):
Some C++-level comments, leaving the semantic check to JIT folks

ezyang commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -912,7 +923,69 @@ void testControlFlow() {
+  return c
+)JIT";
+
+const static auto if_one_expected = R"JIT(  %2 = aten::lt(a, b)
```
Can you use the expect tests capability from Python instead of doing these tests as hardcoded literals? This will make it a lot easier to update the tests if pretty printing changes, as you will be able to just `--accept` the changes.

zdevito commented:
I don't have time for a detailed review right now, but a high-level comment:  When storing meta-data for a particular pass like pretty printing, we do not modify structures in ir.h to hold it but instead create std::unordered_maps (e.g. `std::unordred_map<Value*, std::string> original_name`) that are local to the pass. This keeps implementation details of passes outside of the IR which is shared by many passes.

Commit:
Address comments / refactor

Summary:
Fixed nits, moved tests out to `test_jit.py` and exposed functionality
to Python via `graph.pretty_print()`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented on test/expect/TestScript.test_pretty_printer-while_if_test.expect:
```diff
@@ -0,0 +1,15 @@
+  %5 = aten::lt(a, 10)
+  %6 = prim::TensorToNum(%5)
+  while %6
```
no colon here?

ezyang commented on test/expect/TestScript.test_pretty_printer-if_one.expect:
```diff
@@ -0,0 +1,7 @@
+  %2 = aten::lt(a, b)
```
Not related to this patch, but the mixing of `%2` and `a` really makes my eyes bleed haha

driazati replied:
We could instead use some kind of generated temporary names (e.g. `t1`, `t2`, ...) for this or just remove the intermediate values altogether with something like `prim::TensorToNum(aten::lt(a, b))` but that could get long.

ezyang commented on torch/csrc/jit/ir.cpp:
```diff
@@ -96,6 +96,64 @@ std::string escapeString(std::string s) {
   return s;
 }
 
+void printAttributeValue(std::ostream & out, Symbol & name, const Node * n) {
```
Did the code move only, or are there changes?

driazati replied:
Just a move, this was needed in 2 places

ezyang commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +254,220 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+
+std::ostream& Graph::prettyPrint(std::ostream & out) {
+  constant_list empty_constants;
+  alias_list empty_aliases;
```
Can we have a `Note [Pretty-printing aliases]` explaining what the aliases do?

ezyang replied:
Also, you're passing `empty_aliases` in by mutable reference... so it's not going to be empty after your function is done with it is it?

ezyang commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +254,220 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+
+  indent(out, 1);
+  out << "return ";
+  prettyPrintValue(out, block()->return_node()->inputs()[0], empty_constants, empty_aliases);
```
This looks wrong. Surely there can be more than one return output?

ezyang commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +254,220 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+}
+
+// Handles iterating over nodes/subblocks
+std::ostream& prettyPrintNodeVisitor(
```
This isn't really a visitor, not in the OOP nomenclature, anyway.

ezyang commented on torch/csrc/jit/ir.cpp:
```diff
@@ -250,6 +254,220 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  int count = 1;
+  for (const auto* block_param : body_block->param_node()->outputs())  {
+    // add names to map
+    aliases[block_param->uniqueName()] = node->inputs()[count]->readableName();
```
So... what happens if the loop carried dependency is written before all uses of the loop input are used? I'm thinking about the non-pretty-printed case of:

```
While(cond, carry_init)
  graph (carry_in) {
    carry_out = Op(carry_in)
    new_cond = Op2(carry_in)
    return (new_cond, carry_out)
  }
```

Won't your aliasing logic turn this into:

```
carry_init = ...
while cond:
  carry_init = Op(carry_init)
  cond = Op2(carry_init)
```

That looks bad.

driazati replied:
I think the situation you're describing can't actually happen, the aliases lookup is only for getting a name for a block input from the loop inputs. With the SSA in place, any write will give it a new output name that wouldn't be found in the aliases map. I'm still not entirely convinced about this myself though, are you able to come up with a small concrete example that demonstrates this?

ezyang replied:
Are you saying that the frontend will never generate core in this order? But don't you agree that it can generate:

```
While(cond, carry_init)
  graph (carry_in) {
    new_cond = Op2(carry_in) // now first
    carry_out = Op(carry_in) // now second
    return (new_cond, carry_out)
  }
```

from 

```
while cond:
  cond = carry_init * 2
  carry_init = carry_init * 2
```

And don't you agree that it is a valid optimization to reorder these two instructions? Then the bad graph can arise. Isn't it a design goal for the pretty-printer to work on optimized graphs?

driazati replied:
Ah, I hadn't considered re-ordering, I see what you're saying. What do you think about introducing a temporary if this situation is detected? For example for the script:

```
def while_test(a):
  b = a
  while a < 10:
    b = a * 2
    a = a + 2
  b = a + b
  return b
```

The IR could potentially look like:
```
prim::Loop(cond, a, b)
  block(a_in, b_in) {
    a_out = aten::add(a_in, 2)
    b_out = aten::mul(a_in, 2)
    new_cond = aten::lt(a_out, 10)
    ...
    -> (new_cond, a_out, b_out)
  }
```

And pretty printed:
```
t1 = a 			# added by pretty printer
while cond:
  t1 = a 		# added by pretty printer
  a = aten::add(t1, 2, 1)
  b = aten::mul(t1, 2)
  cond = aten::lt(a, 10)
return b
```

ezyang reviewed (changes requested):
I think there are some soundness issues.

Commit:
Address (most of) ezyang's comments

Summary:
Changed intermediate names to prefix with "t" instead of "%"

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Refactor / fix correctness issues

Summary:
Output is more verbose / looks less nice but is now correct
Statements are generated surrounding loops to match loop dependencies to
outputs

Test Plan:
New test in test_jit.py, see expected outputs

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +255,322 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+
+  // Re-assign block outputs to inputs for next iteration
+  dualIterator<const Value*>(
+    body_block->param_node()->outputs(),
```
`body_block->inputs()`

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +255,322 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  // Re-assign block outputs to inputs for next iteration
+  dualIterator<const Value*>(
+    body_block->param_node()->outputs(),
+    body_block->return_node()->inputs(),
```
`body_block->outputs()`

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +255,322 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  // Add assign loop-dependencies to the outputs of the prim::Loop
+  dualIterator<const Value*>(
+    node->outputs(),
+    body_block->return_node()->inputs(),
```
Note: you can alias `node->outputs()` to `body_block->inputs()` and then you do not have to print these assignments. This is known to be safe because body_block->inputs() are not live after the block so reusing their names is ok.

zdevito reviewed (commented):
This is looking pretty good. I have a bunch of minor comments to address to clean up the code. It would also be good to follow up with a PR that minimizes the number of assignment statements that need to be outputted in many circumstances by attempting to reuse variable names when possible.

Commit:
Address comments

Summary:
* Moved all functionality to PrettyPrintPass struct, renamed methods
accordingly
* Removed unnecessary assignments of node outputs to loop outputs
* Removed alias for loop condition in case it is read after the loop

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

driazati commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +255,322 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  prettyPrintValue(out, body_block->param_node()->outputs()[0], pass);
+  out << ":";
+  out << std::endl;
+
```
This should be covered by the assignments of `body->inputs()` to `node->inputs()` before the loop and `block->inputs()` to `block->outputs()` at the end of the loop body

michaelsuo commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +255,330 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  }
+};
+
+std::ostream& Graph::prettyPrint(std::ostream & out) {
```
Quick request: can we add a `dumpPretty()` method to `Graph` that invokes this function using `std::out`? That would make it easier to use this function from a debugger.

Commit:
Add Graph::dumpPretty()

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

driazati commented:
@ezyang @zdevito Comments/correctness issues should be all addressed, is this ready to go?

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,335 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    // aliases_[body_block->inputs()[0]] = body_block->inputs()[0]->uniqueName();
+    aliases_[body_block->inputs()[0]] = body_block->inputs()[0];
+
+    // The block outputs are not live after the end of the block, so we can use
```
I don't think this is true, what happens in this case where `x` is the value of the output for loop-carried dependency `z`, but still is used after the loop?

```
def foo(y):
   x = y + 1
   z = 4
   while y < 8:
      y += 1
      z = x
  return x, z
```

I would think the safe values to use for the loop carried dependencies are the `node->outputs()` which cannot be live when entering the loop.

zdevito replied:
> In that particular case z gets aliased to x at the end and it outputs return (x, x)

That is wrong though -- if the loop never executes the value of `z` should be `4` not `x`.

driazati replied:
I see. How about something like this:

```python
  def script(y1):
    x = aten::add(y1, 1, 1)
    z1 = aten::add(x, 5, 1)
    t9 = aten::lt(y1, 8)
    t10 = prim::TensorToNum(t9)
    t11 = t10
    t12 = y1
+   z = z1
    while t11:
      y2 = aten::add(t12, 1, 1)
      t17 = aten::lt(y2, 8)
      t18 = prim::TensorToNum(t17)
      t11 = t18
      t12 = y2
+     z = x
    return (x, z)
```

I think the extra assignments are necessary for correctness. A followup PR could deal with pruning the unnecessary ones / improving the naming.

zdevito commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,335 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+      using_generated_name = true;
+    }
+
+    if (using_generated_name) {
```
Isn't this going to run every time a Value is referenced? I feel the `Value* -> pretty_name` mapping should be cached.

zdevito reviewed (commented):
I think there are still a couple minor issues. Can you take a look and let me know?

Commit:
Cache value names

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Fix loop assignment bug

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Update test output

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

apaszke commented on test/test_jit.py:
```diff
@@ -6577,6 +6577,60 @@ def foo(x):
 
         self.assertEqual(foo(input), input)
 
+    def test_pretty_printer(self):
```
Can we please add a new `TestCase` for this? It's unrelated to all other functionality

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
   return out;
 }
 
+class PrettyPrintPass {
```
Can we move this to a different file? It doesn't need to live in `ir.cpp` where it increases compilation time, and doesn't need to be here.

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+  std::unordered_map<const Value*, std::string> value_names_;
+
+  template<class T>
+  void dualIterator(
```
nit: this is often called `zipWith` (modulo the fact that we don't gather the returns).

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    at::ArrayRef<T> list_a,
+    at::ArrayRef<T> list_b,
+    const size_t offset_a,
+    const size_t offset_b,
```
Those offsets are unnecessary. ArrayRefs support dropping initial `N` elements using `list.slice(N)`.

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    auto it_a = list_a.begin() + offset_a;
+    auto it_b = list_b.begin() + offset_b;
+
+    for (; it_a != list_a.end() && it_b != list_b.end(); ++it_a, ++it_b) {
```
It would be better to assert that the `ArrayRef` slices are of the same length

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -251,6 +256,357 @@ std::ostream& operator<<(std::ostream & out, const Graph & g) {
+    out << "while ";
+    printValue(out, body_block->inputs()[0]);
+    out << ":";
+    out << std::endl;
```
nit: it's better to use `"\n"` instead of `std::endl`, because `endl` forces an IO buffer flush. We don't need it here, so better to avoid it.

Commit:
Addresss apaszke's comments

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Move PrettyPrintPass to passes/pretty_print.h/cpp

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

driazati commented on torch/csrc/jit/attributes.h:
```diff
@@ -201,6 +201,89 @@ struct Attributes {
     return get<TensorsAttr>(name);
   }
 
+  template<typename T>
```
This is moved from `ir.cpp` so both it and `pretty_print_pass.cpp` could use it

zdevito commented on torch/csrc/jit/passes/pretty_print.cpp:
```diff
@@ -0,0 +1,335 @@
+    aliases_[body_block->inputs()[0]] = body_block->inputs()[0];
+
+    // Add temporaries for loop-carried dependencies
+    zipWith<const Value*>(
```
Discussed in person: This should be possible to do by using only the `node->outputs()` names to represent the loop-carried dependencies and aliasing the `block->inputs()` names to be the `node->outputs()` name in the loop:

```
x1 = 6
y1 = 5
t = 4

x,y = Loop(x1, y1) 
         block(x2, y2)
         -> (t, x2)        
            

current:
            
  x2 = x1
  y2 = y1
  x = x2
  y = y2
  while ...:
    <body>
    x2 = t
    y2 = x2
    x = x2
    y = y2
  
  
x = x1
y = y1  
while ...:
    <body>
    x = t
    y = x2
```

Commit:
Alias block inputs to node outputs

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Cleanup

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Fix missing condition

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

driazati commented:
@pytorchbot retest this please

Commit:
Add bool casts to test

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito reviewed (approved):
This looks good now!

Commit:
Fix rebase

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ailzhang commented:
@pytorchbot retest this please
